cpp/src/arrow/util/ubsan.h

```diff
@@ -49,5 +49,21 @@ inline T* MakeNonNull(T* maybe_null) {
   return reinterpret_cast<T*>(&internal::non_null_filler);
 }
 
+template <typename T>
```

Comment:
I'm no C++ guru, but can't you make a single method for this by templating the input type too, e.g.

```
template <typename T, typename I = uint8_t>
inline typename std::enable_if<std::is_integral<T>::value, T>::type SafeLoad(
    const I* unaligned) {
  typename std::remove_const<T>::type ret;
  std::memcpy(&ret, unaligned, sizeof(T));
  return ret;
}
```

Reply:
This is correct, I think I like how the methods are now though since it is more explicit when casting and a pass-through when not.  One method could certainly delegate to the other.